File: chapter4code/Pong/Pong/Paddle.js
// Paddle.js
var Paddle = Class.create(Rectangle, {
    initialize: function($super){
        $super();
        this.height = 100;
        this.width = 20;
        this.color = "0011FF";
        this.isPlayerOne = true;
    },
    setPosition: function(x, y){
        this.x = x;
        this.y = y;
    },
    getPosition: function(){
        return {
            'x': this.x,
            'y': this.y
        }
    },
    setIsPlayerOne: function(bool){
        this.isPlayerOne = bool;
    },
    checkInput: function(){
        if (this.isPlayerOne == false) {
            if (myInput.isKeyDown(Input.KEY_UP)) {
                if (this.y > 0) {
                    this.y = this.y - 10;
                }
            }
            else 
                if (myInput.isKeyDown(Input.KEY_DOWN)) {
                    // x,y are taken from the left corner
                    if (this.y < game_height - this.height) 
                        this.y = this.y + 10;
                }
            
        }
        else {
            if (myInput.isKeyDown(65)) { // 'A'
                if (this.y > 0) {
                    this.y = this.y - 10;
                }
            }
            else 
                if (myInput.isKeyDown(90)) { // 'Z'
                    // x,y are taken from the left corner
                    if (this.y < game_height - this.height) 
                        this.y = this.y + 10;
                }
        }
    },
    update: function(){
        this.checkInput();
    }
});

File: dartpong/Paddle.dart
#library('paddle');
#import('dart:html');

class Paddle {
  CanvasRenderingContext2D ctx;
  num height = 100;
  num width = 20;
  var color = "0011FF";
  bool isPlayerOne = true;
  num x,y;
  
  Paddle(this.ctx);
  
  void setPosition(x, y) {
    this.x = x;
    this.y = y;
  }
  
  getPosition() {
    return {'x':this.x, 'y':this.y};
  }
  
  void checkInput() {
    document.on.keyDown.add((KeyboardEvent event) {
      event.preventDefault();
      if (this.isPlayerOne == true) {
        if (event.keyCode == 87) {         // w key
          if (this.y > 0)
            this.y -= 0.05;
        } else if(event.keyCode == 83) {  // s key
          if (this.y < 400 - this.height) 
            this.y += 0.05;
        }
      } else {
        if (event.keyCode == 79) {         // o key 
          if (this.y > 0)
            this.y -= 0.05;
        } else if(event.keyCode == 76) {  // l key
          if (this.y < 400 - this.height) 
            this.y += 0.05;
        }
      }
    }, false);
  }
  
  void update() {
    this.checkInput();
  }
  
  void draw() {
    ctx.save();
    ctx.beginPath();
    ctx.fillRect(this.x, this.y, this.width, this.height);
    ctx.restore();
  }
  
}
